### ISSUE TITLE
Taking Almost 4 Hours+ to Compile !!!

### ISSUE THREAD

**Issue by origamiofficial** [openssl/openssl#19144]
Whenever I try to compile openssl for a Dockers container it takes 4hours+ even on Github actions. Is there any way to speed up the process? I can't find any solution.
I'm compiling for ```linux/amd64, linux/arm64, linux/arm/v7, linux/arm/v6``` platforms.

**Command I'm Using***
```bash
set -e -x && \
    git clone https://github.com/openssl/openssl.git && \
    cd openssl && \
    ./config \
      --prefix=/opt/openssl \
      --openssldir=/opt/openssl \
      no-weak-ssl-ciphers \
      no-ssl3 \
      no-shared \
      -DOPENSSL_NO_HEARTBEATS \
      -fstack-protector-strong && \
    make depend && \
    make && \
    make install_sw && \
    rm -rf \
        /tmp/* \
        /var/tmp/* \
        /var/lib/apt/lists/*
```

**Additional information**
The Dockerfile I'm using to compile:
https://github.com/origamiofficial/docker-pihole-unbound/blob/main/Dockerfile

Github Action Output:
https://github.com/origamiofficial/docker-pihole-unbound/actions/runs/2993395643

**Comment by t8m** [comment]
What about using parallel make if you have multiple cores? I.e., use `make -j8` or so. Also if you're compiling multiple times it might be useful to use ccache.

IMO there is not much that can be done to speed things up. OpenSSL is pretty complex software so it takes many CPU cycles to compile - especially with modern compilers that try to optimize things as much as possible.

**Comment by levitte** [comment]
I might add that `make depend` is unnecessary.

But yeah, it's a big piece of software, that's absolutely true.  I also suspect that ARM processors are a bottleneck in this regard.

**Comment by origamiofficial** [comment]
> What about using parallel make if you have multiple cores? I.e., use `make -j8` or so. Also if you're compiling multiple times it might be useful to use ccache.
> 
> IMO there is not much that can be done to speed things up. OpenSSL is pretty complex software so it takes many CPU cycles to compile - especially with modern compilers that try to optimize things as much as possible.

Now I'm using GitHub Action to compile because in my Core i3 laptop it takes 7-8hours. So using `make -j8` will be helpful for GitHub Actions ?

**Comment by origamiofficial** [comment]
> I might add that `make depend` is unnecessary.
> 
> But yeah, it's a big piece of software, that's absolutely true. I also suspect that ARM processors are a bottleneck in this regard.

Yeah you're right amd64 takes only 1hour but arm builds takes a lot of time !!

**Comment by levitte** [comment]
... this takes me back to yesteryears (the first half of the 1990's), when gcc took about a day to compile on a medium Sun machine of that time.  Some friends became upset when the FSF released a patch to their previous release, on the same day as that release.

**Comment by mattcaswell** [comment]
4 hours seems excessive. I don't really understand why it would take so long. As a typical example, this build under Github actions took 7 mins 8 secs to complete:

https://github.com/openssl/openssl/runs/8225311445?check_suite_focus=true

Granted you are targeting different platforms...but still...

**Comment by beldmit** [comment]
Don't you have enabled, say, link time optimization? It can consume much time...

**Comment by origamiofficial** [comment]
> 4 hours seems excessive. I don't really understand why it would take so long. As a typical example, this build under Github actions took 7 mins 8 secs to complete:
> 
> https://github.com/openssl/openssl/runs/8225311445?check_suite_focus=true
> 
> Granted you are targeting different platforms...but still...

ARM Platforms takes too much time !!

**Comment by origamiofficial** [comment]
> Don't you have enabled, say, link time optimization? It can consume much time...

Sorry, I don't understand what are you talking about. Can you explain?

**Comment by origamiofficial** [comment]
> Don't you have a flag -enable-lto?
> 
> Anyway if you have an AMD64 machine, it's worth considering a cross-compile the build for ARM

As you can see in my docker file there is no such flag. Can you tell me where should I use that flag? Also I've zero knowledge about cross-compile. Any help would be great.
My Dockerfile:
https://github.com/origamiofficial/docker-pihole-unbound/blob/main/Dockerfile

**Comment by beldmit** [comment]
Examples from https://github.com/openssl/openssl/blob/master/.github/workflows/cross-compiles.yml may help

**Comment by t8m** [comment]
Also as I proposed above, if your build machine has multiple CPU cores make -j<number> can greatly speed things up.

**Comment by origamiofficial** [comment]
By using ```nproc | xargs -I % make -j%``` compiling time decreased to 3 hours. Any more suggestions ?

**Comment by delasy** [comment]
https://github.com/thelang-io/cli/actions/runs/3651893662/jobs/6169644046
It takes two minutes in GitHub Actions to build with `no-tests` for amd64, try adding `no-tests` and see if it helps. Though I'm not sure if OpenSSL team recommends doing so.

afaik Docker by default is not using full potential of machine it runs on.
For example: machine could be 4GB RAM, 4vCPU and Docker is using only 500MB RAM and 1vCPU
Tweaking docker settings could decrease compilation time.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the end of the 3.4 dev cycle barring further input
